File: Chapter 2 - Linked Lists/2.6_Palindrome/Stack.java
import java.util.EmptyStackException;

public class Stack {
    public class Node {
        public Node next = null;
        public int value;
        public Node(int value) {
            this.value = value;
        }
    }
    public Node head;

    public void push(int value) {
        if (head == null) {
            head = new Node(value);
        } else {
            Node newNode = new Node(value);
            newNode.next = head;
            head = newNode;
        }
    }

    public int pop() {
        if (head == null) {
            throw new EmptyStackException();
        }
        Node oldNode = head;
        head = head.next;
        return oldNode.value;
    }

    public int peek() {
        if (head == null) {
            throw new EmptyStackException();
        }
        return head.value;
    }

    public static void main(String[] args) {
        Stack list = new Stack();
        list.push(5);
        list.push(4);
        list.push(3);
        for (int i=0; i<3; i++) {
            System.out.println(list.pop());
        }
    }
}

File: Python/Chapter 03 - Stacks and Queues/stack.py
class StackNode:
	def __init__(self, value, next):
		self.value = value
		self.next = next

class Stack:
	def __init__(self):
		self.head = None

	def pop(self):
		if self.head is None:
			print("Empty stack")
			return
		value = self.head.value
		self.head = self.head.next
		return value

	def push(self, value):
		self.head = StackNode(value, self.head)

	def peek(self):
		if self.head is None:
			print("Empty stack")
			return
		return self.head.value

	def isEmpty(self):
		return self.head is None

stack = Stack()
stack.push(1)
stack.push(2)
stack.push(3)
print(stack.peek())
print(stack.pop())
print(stack.pop())
print(stack.pop())
print(stack.pop())
